semantic-tags/src/AST/Element.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE AllowAmbiguousTypes, DataKinds, FlexibleInstances, MultiParamTypeClasses, PatternSynonyms, ScopedTypeVariables, TypeApplications, TypeFamilies, TypeOperators, UndecidableInstances, ViewPatterns #-}
+module AST.Element
+( Element(..)
+, pattern Prj
+, (:+:)(..)
+) where
+
+import GHC.Generics ((:+:)(..))
+import GHC.TypeLits (ErrorMessage(..), TypeError)
+
+-- | Projection of an element out of a tree of sums, particularly suitable for use with highly branching (e.g. balanced) trees. This provides custom type errors when the tree is known statically, rather than provided via a constraint passed on to callers.
+class Element sub sup where
+  -- | Project one element out of a sum type.
+  prj :: sup a -> Maybe (sub a)
+
+instance (Element' side sub sup, side ~ Find sub sup) => Element sub sup where
+  prj = prj' @side
+
+
+-- | A pattern synonym to conveniently project out matching elements.
+pattern Prj :: Element sub sup => sub a -> sup a
+pattern Prj sub <- (prj -> Just sub)
+
+
+-- | Where does the element occur in the tree?
+data Side = None | Here | L | R
+
+-- | Find where in a tree an element occurs.
+type family Find sub sup :: Side where
+  Find t t         = 'Here
+  Find t (l :+: r) = Find' 'L t l <> Find' 'R t r
+  Find _ _         = 'None
```

Comment:
I opted to use a (recursive 😱) type family for this because:

1. the previous approach, based on reassociating the tree like `fused-effects` does, becomes much more expensive when your trees are balanced rather than generally right-leaning with a few n < 3 branches on the left;
2. we’re not building recursive witnesses (e.g. lists of sides indicating the path to a given element in the tree), so there are fewer type equalities having to be maintained (tho certainly still some);
3. the arities of these trees are generally small given appropriate use of supertypes in the grammars; and
4. it allows us to produce _really good_ type error messages in failure cases:

    ```haskell
    newtype A a = A a deriving Show
    newtype B a = B a deriving Show
    newtype C a = C a deriving Show
    newtype D a = D a deriving Show
    newtype E a = E a deriving Show

    prj (L1 (R1 (B 'b')) :: ((A :+: B) :+: (C :+: D)) Char) :: Maybe (B Char)
    -- => Just (B 'b')

    prj (L1 (R1 (B 'b')) :: ((A :+: B) :+: (C :+: D)) Char) :: Maybe (E Char)
    {- => 
    <interactive>:9:1-55: error:
        • E is not in
          { A
          , B
          , C
          , D
          }
    -}
    ```

    (Yes, I _did_ spend entirely too long getting that to format & align nicely. Why do you ask?)

Reply:
I think this is both a reasonable implementation (it may be a recursive type family, but sometimes functions are recursive, and what is a type family but a function on the type level?) especially given the circumstances (the maximum `N` here is, like, ~15 for now, rather than the 145 we had for DTALC TypeScript.

Reply:
Yeah, the real concern is that recursive type families (can?) build up O(n²) coercions as witnesses of equalities.